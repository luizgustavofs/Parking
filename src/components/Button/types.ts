import { DefaultTheme } from 'styled-components';

export interface ButtonBackgroundProps {
  primary?: boolean;
  color?: string;
  secondary?: boolean;
  disabled?: boolean;
  withOutBG?: boolean;
}

export interface getStyleProps extends ButtonBackgroundProps {
  theme: DefaultTheme;
}

export interface ButtonProps extends ButtonBackgroundProps {
  children: string;
  onClick?: () => void;
}
``;
