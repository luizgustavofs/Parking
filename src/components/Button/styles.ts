import styled from 'styled-components';

import { ButtonBackgroundProps, getStyleProps } from './types';

const getStyle = ({
  theme,
  color,
  disabled,
  primary,
  secondary,
  withOutBG,
}: getStyleProps) => {
  if (primary) {
    return `
      background-color: ${disabled ? theme.colors.neutral._70 : color}; 
      h4 {
        color: ${disabled ? theme.colors.neutral._80 : theme.colors.neutral._00}
      }; 
      ${disabled && `cursor: auto;`}
    `;
  }
  if (secondary) {
    return `
    border-style: solid; border-width: 2px; border-color: ${
      disabled ? theme.colors.neutral._80 : color
    }; 
    h4 {
      color: ${disabled ? theme.colors.neutral._80 : color}
    }; 
      ${disabled && `cursor: auto;`}
    `;
  }
  if (withOutBG) {
    return ` 
    background: #fff;
    h4 {
      color: ${disabled ? theme.colors.neutral._80 : color}
    };
    ${disabled && `cursor: auto;`}
  `;
  }
};

export const Container = styled.button<ButtonBackgroundProps>`
  display: flex;
  width: 100%;
  height: 4.188rem;
  border: none;
  border-radius: 4px;

  cursor: pointer;

  align-items: center;
  justify-content: center;
  margin-bottom: 16px;

  ${props => getStyle(props)};
`;
