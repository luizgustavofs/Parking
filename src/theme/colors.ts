export const colors = {
  default: '#000000',
  primary: '#4DD0E1',
  secondary: '#F2F2F2',
  tertiary: '#28DD91',
  quaternary: '#A769B2',
  error: '#FF0068',
  baseColor: '#FFFBE6',
  neutral: {
    _00: '#ffffff',
    _70: '#DADADA',
    _80: '#9B9B9B',
  },
};
