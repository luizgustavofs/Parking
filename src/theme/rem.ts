export const REM_BASE = 16;
