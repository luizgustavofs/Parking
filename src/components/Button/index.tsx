import React from 'react';

import { useTheme } from 'styled-components';

import { Container } from './styles';
import { ButtonProps } from './types';

const Button: React.FC<ButtonProps> = ({
  children,
  color,
  primary,
  secondary,
  withOutBG,
  disabled,
  onClick,
}) => {
  const { colors } = useTheme();

  return (
    <Container
      onClick={onClick}
      withOutBG={withOutBG}
      disabled={disabled}
      primary={primary}
      secondary={secondary}
      color={color || colors.tertiary}>
      <h4>{children}</h4>
    </Container>
  );
};

export default Button;
