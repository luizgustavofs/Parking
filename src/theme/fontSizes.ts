import { REM_BASE } from './rem';

export const fontSizes = {
  h1: 38 / REM_BASE,
  h2: 24 / REM_BASE,
  h3: 16 / REM_BASE,
  h4: 15 / REM_BASE,
  h5: 12 / REM_BASE,
};
