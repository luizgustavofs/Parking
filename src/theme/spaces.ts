import { REM_BASE } from './rem';

export const spaces = {
  x1: 4 / REM_BASE,
  x2: 8 / REM_BASE,
  x3: 12 / REM_BASE,
  x4: 16 / REM_BASE,
  x6: 24 / REM_BASE,
  x8: 32 / REM_BASE,
  x10: 40 / REM_BASE,
  x11: 44 / REM_BASE,
  x12: 48 / REM_BASE,
  x13: 52 / REM_BASE,
  x14: 56 / REM_BASE,
  x16: 64 / REM_BASE,
  x18: 72 / REM_BASE,
  x20: 80 / REM_BASE,
  x22: 88 / REM_BASE,
  x24: 96 / REM_BASE,
  x30: 120 / REM_BASE,
  x40: 160 / REM_BASE,
  x50: 200 / REM_BASE,
};
