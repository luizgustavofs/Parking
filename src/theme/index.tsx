import { ReactNode } from 'react';

import { ThemeProvider } from 'styled-components';

import { colors } from './colors';
import { fontSizes } from './fontSizes';
import { spaces } from './spaces';

export const theme = {
  colors,
  spaces,
  fontSizes,
};

const ProjectThemeProvider: React.FC<{ children: ReactNode }> = ({
  children,
  ...props
}) => {
  return (
    <ThemeProvider theme={theme} {...props}>
      {children}
    </ThemeProvider>
  );
};

export { ProjectThemeProvider };
