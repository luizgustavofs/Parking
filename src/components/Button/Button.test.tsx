import { fireEvent } from '@testing-library/dom';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import Button from '.';
import { theme } from '../../theme';

describe('Button', () => {
  it('should render correctly with variant primary', () => {
    const { toJSON } = create(
      <ThemeProvider theme={theme}>
        <Button primary>test</Button>
      </ThemeProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render correctly with variant secondary', () => {
    const { toJSON } = create(
      <ThemeProvider theme={theme}>
        <Button secondary>test</Button>
      </ThemeProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should call onClick prop when clicked', () => {
    const buttonTestId = 'button';
    const mockFunction = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button onClick={mockFunction} testID={buttonTestId}>
          test
        </Button>
      </ThemeProvider>,
    );

    const button = getByTestId(buttonTestId);
    fireEvent.click(button);
    expect(mockFunction).toHaveBeenCalled();
  });

  it('should not call onClick when disabled', () => {
    const buttonTestId = 'button';
    const mockFunction = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button disabled onClick={mockFunction} testID={buttonTestId}>
          test
        </Button>
      </ThemeProvider>,
    );

    const button = getByTestId(buttonTestId);
    fireEvent.click(button);
    expect(mockFunction).not.toHaveBeenCalled();
  });

  it('should change color when its passed by props', () => {
    const buttonTestId = 'button';
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          primary
          color="#3a3bbb"
          onClick={() => {}}
          testID={buttonTestId}>
          test
        </Button>
      </ThemeProvider>,
    );

    const button = getByTestId(buttonTestId);
    expect(button).toHaveStyle({ backgroundColor: '#3a3bbb' });
  });

  it('should ignore the color props when disabled and show with disabled color', () => {
    const buttonTestId = 'button';
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          primary
          color="#3a3bbb"
          disabled
          onClick={() => {}}
          testID={buttonTestId}>
          test
        </Button>
      </ThemeProvider>,
    );

    const button = getByTestId(buttonTestId);
    expect(button).toHaveStyle({ backgroundColor: theme.colors.neutral._70 });
  });

  it('should not have background color with secondary variant', () => {
    const buttonTestId = 'button';
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          secondary
          color="#3a3bbb"
          onClick={() => {}}
          testID={buttonTestId}>
          test
        </Button>
      </ThemeProvider>,
    );

    const button = getByTestId(buttonTestId);
    expect(button).toHaveStyle({ backgroundColor: 'none' });
  });

  it('should not have background and border color with prop withOutBG', () => {
    const buttonTestId = 'button';
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          withOutBG
          color="#3a3bbb"
          onClick={() => {}}
          testID={buttonTestId}>
          test
        </Button>
      </ThemeProvider>,
    );

    const button = getByTestId(buttonTestId);
    expect(button).toHaveStyle({
      backgroundColor: 'none',
      borderColor: 'none',
    });
  });
});
