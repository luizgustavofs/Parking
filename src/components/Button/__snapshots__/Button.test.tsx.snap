// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly with variant primary 1`] = `
<button
  className="sc-bczRLJ bIDAMy"
  color="#28DD91"
>
  <h4>
    test
  </h4>
</button>
`;

exports[`Button should render correctly with variant secondary 1`] = `
<button
  className="sc-bczRLJ jLgYQE"
  color="#28DD91"
>
  <h4>
    test
  </h4>
</button>
`;
